### PR TITLE
Improves reference search by StructLiteral

### PR DESCRIPTION
**Pull request by mikhail-m1**
Hey, I've made some changes to improve search for struct literals, now it works for `struct Foo<|> {`, `struct Foo <|>{`, `struct Foo<|>(`. Unfortunately tuple creation is represented as a call expression, so for tuples it works only is search is started in a tuple declaration.  It leads to incorrect classification of function calls during search phase, but from user perspective it's not visible and works as expected. May be it worth to add a comment or rename it to remove this misleading classification. Issue #2549.

### CODE REVIEW

File: crates/ra_ide/src/references.rs [hunk 1]
```diff
@@ -122,8 +122,33 @@ pub(crate) fn find_all_refs(
     let mut search_kind = ReferenceKind::Other;
 
     if let TokenAtOffset::Between(ref left, ref right) = token {
-        if (right.kind() == SyntaxKind::L_CURLY || right.kind() == SyntaxKind::L_PAREN)
-            && left.kind() != SyntaxKind::IDENT
+        let opt_name = find_node_at_offset::<ast::Name>(&syntax, left.text_range().start());
+        let opt_name_ref = find_node_at_offset::<ast::NameRef>(&syntax, left.text_range().start());
```
Comment (kjeremy) [inline crates/ra_ide/src/references.rs]: Can both of these be `Some` at the same time?

**Comment by kjeremy** [comment]
That's a fairly complicated if condition. It may be clearer if you can split it up into distinct chunks one per case and see what falls out of that.

**Comment by mikhail-m1** [comment]
> That's a fairly complicated if condition. It may be clearer if you can split it up into distinct chunks one per case and see what falls out of that.

The condition requires some refactoring, I think helper functions would help,  and    `process_definition` should be refactored as well.

File: crates/ra_ide/src/references.rs [hunk 2]
```diff
@@ -122,8 +122,33 @@ pub(crate) fn find_all_refs(
...
+                        .and_then(|p| p.segment())
+                        .map(|p| p.name_ref() == opt_name_ref)
+                        .unwrap_or(false))))
+            || (right.kind() == SyntaxKind::L_PAREN // StructDef<|>(
+                && opt_name.is_some()
+                && left
+                    .ancestors()
+                    .find_map(ast::StructDef::cast)
+                    .map(|l| l.name() == opt_name)
+                    .unwrap_or(false))
```
Comment (matklad) [inline crates/ra_ide/src/references.rs]: Yeah, I guess it's a strong contender for the biggest expression I've ever seen in an `if` condition. Moving it to a helper function with some early returns would help here I think!

File: crates/ra_ide/src/references.rs [hunk 3]
```diff
@@ -122,8 +123,12 @@ pub(crate) fn find_all_refs(
     let mut search_kind = ReferenceKind::Other;
 
     if let TokenAtOffset::Between(ref left, ref right) = token {
-        if (right.kind() == SyntaxKind::L_CURLY || right.kind() == SyntaxKind::L_PAREN)
-            && left.kind() != SyntaxKind::IDENT
+        if (((right.kind() == SyntaxKind::WHITESPACE
+            && right.next_token().map(|t| t.kind()) == Some(SyntaxKind::L_CURLY))
+            || right.kind() == SyntaxKind::L_CURLY)
```
Comment (edwin0cheng) [inline crates/ra_ide/src/references.rs]: What happens if we have two white space tokens here ?

Comment (kiljacken) [reply crates/ra_ide/src/references.rs]: There should never be more than one `WHITESPACE` token in a row, so this seems reasonable (although the if condition is a bit large).

File: crates/ra_ide/src/references.rs [hunk 4]
```diff
@@ -322,7 +356,42 @@ mod tests {
...
             "Foo STRUCT_DEF FileId(1) [5; 39) [12; 15) Other",
-            &["FileId(1) [142; 145) StructLiteral"],
+            &["FileId(1) [138; 141) StructLiteral"],
+        );
+    }
+
+    #[test]
+    fn test_struct_literal_befor_space() {
+        let code = r#"
+    struct Foo<|> {}
```
[the remark this answers is not in the document]
Comment (mikhail-m1) [reply crates/ra_ide/src/references.rs]: @matklad  I decided to create workaround after https://github.com/rust-analyzer/rust-analyzer/pull/2738#issuecomment-572075163 for now, but I can restore previous behavior.  Also I've noticed that it doesn't work for `struct Foo<T> {`, should it work for `struct Foo<|><T> {` or `struct Foo<T> <|>{` ?

[the remark this answers is not in the document]
Comment (mikhail-m1) [reply crates/ra_ide/src/references.rs]: actually `struct Foo <|>{` works for me in vscode, so I will roll back `struct Foo<|> {`

**Comment by matklad** [comment]
It was fixed like 40 minutes ago:
https://github.com/microsoft/vscode/issues/88413#event-2954179485 :)

**Comment by mikhail-m1** [comment]
> It was fixed like 40 minutes ago: [microsoft/vscode#88413 (comment)](https://github.com/microsoft/vscode/issues/88413#event-2954179485) :)

I definitely tested before the it, might be related to vim extension.

**Comment by flodiebold** [comment]
I think this needs a rebase before bors can merge it.

**Comment by matklad** [comment]
bors d+

**Comment by mikhail-m1** [comment]
I rebased it and pushed again, but was no conflicts so it's weird.
bors r+

**Comment by matklad** [comment]
it's not about conflict, its about bors acting weird if there's a change to
`.github/workflows` in the diff between PR's HEAD and master

**Pull request merged**